src/charm.py

```diff
@@ -785,12 +789,15 @@ def _web_config(self) -> Optional[dict]:
         Ref: https://prometheus.io/docs/prometheus/latest/configuration/https/
         """
         if self._is_tls_enabled():
-            return {
-                "tls_server_config": {
-                    "cert_file": CERT_PATH,
-                    "key_file": KEY_PATH,
+            if not self.container.exists(CERT_PATH):
+                self.stop()
```

Comment:
Seems odd to issue a stop from a config getter func.
Is there anywhere else this could fit better?

Reply:
I don't think so; this is also the approach that's already used in grafana-agent :)

Reply:
At the very least we should add a "side effect" section to the docstring.

But this change seems to deepen the spaghetti: we have a `stop` here but not a `start`.
If this is the best we can do then maybe add a comment here?
```
# After a `stop`, the service will autostart on next call to `_configure`, which is expected
# to happen as soon as the the related CA replies with a cert.
```

Iiuc, there's a complication here with relation-broken: on relation-broken I imagine `cert_handler.enabled` would still return True, but what happens with CERT_PATH? Is it deleted before or after? And maybe for this reason we should move the `stop` elsewhere?

Reply:
The reason why `stop` is there is because that's the point in the code where we decide if we have TLS or not (and thus create the web config and pass it via the Pebble plan).  
The fact that the service is reloaded-restarted in `_configure`, is already documented in configure; sure I'll add that extra comment, but it's not new behavior (nor particularly quirky; lots of our charms do this). We don't need `start` because Pebble `replan()` already does this.  

`cert_handler.enabled` is currently implemented as:
```python
    @property
    def enabled(self) -> bool:
        """Boolean indicating whether the charm has a tls_certificates relation."""
        # We need to check for units as a temporary workaround because of https://bugs.launchpad.net/juju/+bug/2024583
        # This could in theory not work correctly on scale down to 0 but it is necessary for the moment.
        return (
            len(self.charm.model.relations[self.certificates_relation_name]) > 0
            and len(self.charm.model.get_relation(self.certificates_relation_name).units) > 0  # type: ignore
        )
```

About `CERT_PATH` being deleted before or after relation-broken, it doesn't matter; `cert_handler.enabled` will return false; the charm will still exclude the TLS configuration because the relation is not there. The relation-broken issues would be better addressed in [cert_handler:_on_server_cert_changed()](https://github.com/canonical/prometheus-k8s-operator/blob/main/src/charm.py#L386), where we could check if we are in a relation-broken event and eventually just remove the certs.    

**Even if** `cert_handler.enabled` returned true, the issue concerns `cert_handler` and not the individual charm, and as such it's not part of the scope of this issue.

I'm fully open to reconsider if methods like `_configure` should have side effects like stopping the charm, however I want to reiterate [we already do this in grafana-agent](https://github.com/canonical/grafana-agent-k8s-operator/blob/main/src/grafana_agent.py#L479); I'd leave further design discussions out of this issue.

Reply:
We decided that we should move these checks to the outer `_configure ` method.

Reply:
Also, we should change to `WaitingStatus` while Prometheus is stopped.